### ISSUE TITLE
Server side `.render()`?

### ISSUE THREAD

**Issue by olalonde** [facebook/react#6964]
I was wondering if it was possible to render a react component server side without serialising it to a string. It should make it possible to trigger state changes without re-rendering the whole tree.

For example, I'd like to be able to do something like this (pseudo-code):

```javascript
class Test extends Component {
  componentWillMount() {
    // this method should only be called once, by `render(Test)`
    setTimeout(() => this.setState({ ready: true }), 1000)
  }
  render() {
    const { ready } = this.state
    if (ready) {
      return (<strong>I am ready</strong>)
    }
    return  (<strong>I am not ready</strong>)
  }
}

const tree = render(Test)
renderToString(tree)
// => <strong>I am not ready</strong>
setTimeout(() => renderToString(tree), 2000)
// => <strong>I am ready</strong>
```

I don't really plan to use timeouts but it was to illustrate the idea.

This would be another way to put it:

```javascript
let setState
class Test extends Component {
  componentWillMount() {
    // important: this is called only once
    console.log('componentWillMount called')
    setState = this.setState.bind(this)
  }
  render() {
    const { ready } = this.state
    if (ready) {
      return (<strong>I am ready</strong>)
    }
    return  (<strong>I am not ready</strong>)
  }
}

const tree = render(Test)
// outputs: componentWillMount called
console.log(renderToString(tree))
// outputs: <strong>I am not ready</strong>
setState({ ready: true })
console.log(renderToString(tree))
// outputs: <strong>I am ready</strong>
```

If this is not possible currently, would it be difficult to implement?

**Comment by Aweary** [comment]
What's your use case here? Why do you want to render a component multiple times and update state on the server?

**Comment by jimfb** [comment]
I assume @olalonde is trying to have components load data by calling setState when the data arrives.

There are a couple problems with this approach, and it is generally not a great pattern.
 * On the client side, this means that you can't start loading data for a child component until the parent component has finished rendering with data, which means that a component tree of depth 20 will require 20 round trips to the server to fetch all the data and have the page load be complete.  It would be much better to figure out what data is needed at the top, and then fetch everything in a single trip at the beginning.
 * On the server side, you eventually need to send markup down to the client.  But there is no way of knowing when a render is "done" and thus "ready to be sent to a client", because there could be outstanding data requests that could call setState on your component at any point in the future.
 * The component data is not internal component state.  In my experience, using `setState` (or any other mechanism) to have stateful components makes your code much more difficult to reason about.  But in this particular case, it's especially unnecessary.  If the component's data just a copy of sideways loaded data, you are duplicating the source of truth.  Related reading: https://facebook.github.io/react/tips/props-in-getInitialState-as-anti-pattern.html

The list goes on, there are a couple other issues, but in the interest of time, I'll truncate the list here.  In general, I recommend NOT fetching data in `component[Will/Did]Mount` and NOT storing data in state.

To the extent that this is a request for async SSR data, it's a duplicate of https://github.com/facebook/react/issues/1739

**Comment by mxstbr** [comment]
> NOT storing data in state

Where would you store data you fetched from a server?

**Comment by jimfb** [comment]
@mxstbr In some external [Flux](https://facebook.github.io/flux/docs/overview.html) datastore.  Redux is pretty popular.  We use Relay at facebook.  There is no dearth of data management frameworks at your disposal.  The point is that the data is general application state, not internal component state, so it should be modeled at the application layer.  Modeling the data at the correct layer is the only way to keep your application fast and internally consistent.

**Comment by mxstbr** [comment]
Ah, I thought "NOT storing data in state" included application state, my bad! Thanks for the answer!

**Comment by olalonde** [comment]
@jimfb wow, you read way too much into my example 😆 

1) This is inevitable regardless of React. For example, my app displays the currently logged in user's username in the header so a full load of any given page will have at least 1 + n API requests. Fetching everything in a single round trip is not always possible especially if you have no control over the API you are using.

> which means that a component tree of depth 20 will require 20 round trips to the server to fetch all the data and have the page load be complete

I don't see why this would necessarily be the case. In my case, Most of my pages require 2 round trips: one to get the logged in user profile and another to load the rest of the data. 

> It would be much better to figure out what data is needed at the top, and then fetch everything in a single trip at the beginning.

In terms of efficiency, yes. In terms of maintainability, no.

2) On the server side, you eventually need to send markup down to the client. But there is no way of knowing when a render is "done" and thus "ready to be sent to a client", because there could be outstanding data requests that could call setState on your component at any point in the future.

That is incorrect. I keep track of all async calls done in `componentWillMount` through a counter and know when a render is done (without using timeouts 😄 ). Here's the code I'm using right now: https://gist.github.com/olalonde/f3dc38c8c424c3a96d7a699c57f981fd. It's not really efficient but would be if React offered the API I described in the issue.

3) I am using redux...

**Comment by jimfb** [comment]
> Regardless, I don't think we should have a debate about any of the stuff you mentioned when all I asked for a quite simple API and I went through some effort to avoid mentioning async data loading because I know it is the subject of much controversy and there are quite a lot of issues dedicated to that topic.

I don't understand the motivation for your request, if not for async data loading.  Can you elaborate on use cases other than data loading?

**Comment by olalonde** [comment]
It could be used to test React server side for example. Anyways, I take it it wouldn't be easy to implement which is fine.

**Comment by jimfb** [comment]
> It could be used to test/simulate React server side for example.

I still don't understand.  If your motivation is testing (including updates), why are you using `renderToString` instead of polyfilling the browser API (eg. jsdom, or using a real browser) and just using `ReactDOM.render()`?
